### PR TITLE
Fix fish test "framework"

### PR DESCRIPTION
**Pull request by haampie**
It seems not possible to change the exit code inside an exit trap, so tests were always passing even though they have been failing on develop ~for some time~ since February.

### CODE REVIEW

**Comment by adamjstewart** [comment]
@JBlaschke @pat-s @gonsie can any of you review this?

File: share/spack/qa/setup-env-test.fish [hunk 1]
```diff
@@ -116,7 +118,7 @@ function spt_fails
 
     if test $status -eq 0
         fail
-        echo_red "Command failed with error $status"
+        echo_red "Command succeeded, but should fail"
```
Comment (alalazo) [inline share/spack/qa/setup-env-test.fish:122]: It's not clear to me why the change above is not needed here.

Comment (alalazo) [reply share/spack/qa/setup-env-test.fish:122]: I mean, setting `cmd_status` immediatly after `set -l output (eval $argv 2>&1)`

Comment (haampie) [reply share/spack/qa/setup-env-test.fish:122]: In this case $status is used only once, in the others it's used twice, and then the second time it refers to the status of another command

Comment (haampie) [reply share/spack/qa/setup-env-test.fish:122]: In this case it doesn't make sense to print the status code, since it's testing if the command fails to fail, so it's always 0 anyways.

Comment (alalazo) [reply share/spack/qa/setup-env-test.fish:122]: Oh, I missed that it's not in the string anymore. Can we use `cmd_status` also here, for better clarity if we ever change this file, and add a comment in both cases above the line where we:
```
set cmd_status $status
```
that says we record the status since following command may override it? Other than this minor things this LGTM.

File: share/spack/qa/setup-env-test.fish [hunk 2]
```diff
@@ -90,9 +90,11 @@ function spt_succeeds
 
     set -l output (eval $argv 2>&1)
 
-    if test $status -ne 0
+    set cmd_status $status
```
Comment (JBlaschke) [inline share/spack/qa/setup-env-test.fish:94]: Is this still necessary? Your change on line 257 replacing the trap with the `function -p` should allow `$status` to work again. No?

File: share/spack/qa/setup-env-test.fish [hunk 3]
```diff
@@ -141,10 +143,13 @@ function spt_contains
     printf "'$remaining_args' output contains '$target_string' ... "
 
     set -l output (eval $remaining_args 2>&1)
+    set cmd_status $status
```
Comment (JBlaschke) [inline share/spack/qa/setup-env-test.fish:149]: Same here. Do we still need `cmd_status` after your change on line 257?

**Review by JBlaschke — DISMISSED** [review]
Approving because it fixes a bug. Thanks for spotting that @haampie . I have a suggested change though:

I wrote a `stderr`, `stdout`, and `status` redirection workaround here: https://github.com/spack/spack/blob/264b00bff4220fcb5470e1b80f80e543306ae991/share/spack/setup-env.fish#L157 -- should we use that? It might make things a little clearer and maintainable.

File: share/spack/qa/setup-env-test.fish (hunk 2 above)
Comment (haampie) [reply share/spack/qa/setup-env-test.fish:94]: `$status` changes from command to command:

```fish
#!/usr/bin/env fish

function example
  true
  echo "$status"
  false
  echo "$status"
end

example
```

echos `0` and `1`. So i'm just storing $status to print it later in `echo_red`

File: share/spack/qa/setup-env-test.fish (hunk 1 above)
Comment (haampie) [reply share/spack/qa/setup-env-test.fish:122]: Hm, but the status in `spt_fails` is not really interesting, if and `spt_fails` test fails the status is 0.

File: share/spack/qa/setup-env-test.fish (hunk 2 above)
Comment (JBlaschke) [reply share/spack/qa/setup-env-test.fish:94]: Ah! I see it in the echo statement now!

**Review by JBlaschke — APPROVED** [review]

**Review by alalazo — APPROVED** [review]
